Support KNX event 'GroupValue_Read'

waldbaer opened the pull request:
Support the KNX event type 'GroupValue_Read'.

This is used by devices to trigger update of values (e.g. date and time).
Tested with MDT BE-GT2TW.01 time communication object requested on device startup.

Please let me know if you have any improvements to the code I can implement.

waldbaer commented:
During resolution of the merge conflict I detected that my first approach only worked if the input node is configured to DPT 1.001.
In case of GroupValue_Read the KNX requestor does not provide any value (see also [KNX module docu](https://bitbucket.org/ekarak/knx.js/src/master/README-events.md?fileviewer=file-view-default) ).

Therefore a separate event handler without value parameter is required here.

atlewee commented:
I did some updates today but waited with merging your pull request until I could test it locally. Was planning to look into this tomorrow. If you have a solution please feel free to update your pull request.

waldbaer commented:
Yes, I improved my initial commit (force-commit).
Now I could configure the correct DPT type in the input node and a proper input message gets generated. Payload and rawValue is now 'null' in case of these GroupRead events.

Please let me know if you find any improvement / bug.
I would be happy if I can help improving this very nice node-red plugin. All other KNX modules for node-red do not work very well in my opinion. I prefer using knx-easy!

waldbaer commented:
If you are using knxd you can simply test it using the following command
`$ knxtool groupread ip:localhost 1/0/99`

waldbaer commented:
After I was now able to handle receive GroupValue_Read requests I also wanted to transmit them by myself. This is useful if you want to implement a own KNX visualization using e.g. node-red dashboard. On system startup you could query the actors to get the current state of the lights (for example).

So I extended the knxEasy-out with a `Mode` parameter supporting the values `Read` and ´Write`. Now simply a read() instead of a write() API call is triggered.
The response can simply be read by the knxEasy-in node (no changes required here).

I also thought about introduction of a new node type `knyEasy-read-response` for example having 1 input AND additionally 1 output  where the response is provided. Or the existing knxEasy-out node could be extended with an output port.

What do you think?
Maybe it makes sense to split these 2 commits into 2 PullRequests. Please let me know. I am open to everything :)

Best regards
Sebastian

atlewee commented:
Sorry my late reply, I did not have time to test before today.
I was inspired by your pull request, but the solution is a bit different.
I added the read part to the knx-in node instead of the out-node.

So if you send a message to the input of the Knx-In node, it will do a readRequest for that group address and give you the result on the output of that same node.

I also added a checkbox to the Knx-In node where you can select to automatically send a read request on connection, (so you don't need to create a separate injection node for that)
An added benefit of this is that  it will also send a read request if there is a communication failure and the communication comes back.  So that you do not loose track of changes that happened when the connection was down.
I have released version 0.1.0.  Please try it out..

This was a great suggestion, thanks!
My KNX data is more reliable now.
(I guess this can also be used as a simple communication alarm, just hook up an injection node to repeat a msg evry 10s, and then check with an RBE node if there has been no response in a given time to raise the alarm)

Will this solve your needs for GroupRead?

waldbaer committed:
Support input event "GroupValue_Read"

* Improved knxEasy-in help

waldbaer commented:
Wow, now this is a amazing solution! Especially the feature to automatically send read requests on startup simplifies the usage dramatically. Thanks!

The only thing which is missing for my usecase is the support to handle `GroupValue_Read` requests by the knxEasy-in node.
My KNX push button automatically sends such a read request on startup to query the current date and time. Because I have no weather station or similar I want to handle the request for the current time with knxEasy-in. So when I receive such a `GroupValue_Read` request I query the current time via node-red-contrib-moment and transmit it via knxEasy-out. Then the push button directly has the current time.

I reworked this pull-request and added the event type `GroupValue_Read` to the knxEasy-in node.
Please have a look on my commit 4e9cde8d3da745fff94e19903829d80dbd2c4c5d.

That's the way it makes fun to contribute! Thank you!

atlewee commented:
Ok, I see, you want to be able to respond to external readRequests.
I have not tested this pull request, but I think there is a small problem with the current suggestion:
If there is a input-node that gives values to a consumer, if a read-request is received, the consumer will get a message with payload null ?  This should be avoided.

I see two ways to avoid this:
1) Have a separate node for listening to read requests.
2) Add another output  to the Knx-In node.

What do you think ?

I think I'm in favor of #2 , but it will add a little more complexity to the input node for a beginner to understand it.
But I guess with good documentation, it should not be a problem?

Implementing #2 should be fairly simple..
Don't know if you are familiar with multliple outputs, but its just an array of values, and null means do not send a msg on this output.

if (readRequest) then input.send( [null, readMsg] )
else if (evt is write or response) then input.send( [msg, null] )

waldbaer commented:
You are right, the messages with payload null could be confusing for beginners.
I think adding a second output is a little bit too much.

What do you think about variant 3):
- Add a checkbox to the in node "Notify read requests". Only if this checkbox is enabled, read requests will be notified with the payload value null. Good documentation of this checkbox is mandatory. By default this checkbox should be disabled.

An experienced user can then filter the read requests with a simple switch node: `msg.knx.event == "GroupValue_Read"` or `msg.payload == null`. An example how to do this should be given in the documentation.

What do you think?

waldbaer committed:
Add config option `Notify read requests`

* Notify about received read requests only if option is enabled.
* Adapted in node properties documentation according help style guide (https://nodered.org/docs/creating-nodes/help-style-guide)
